### PR TITLE
[rl] Make ChunkedLoss configurable with a pluggable inner loss (+ GRPO on chunked loss)

### PR DESCRIPTION
**Pull request by DamianSzwichtenberg**
## Summary

This PR refactors the cross-entropy-only `ChunkedCELoss` on `main` into a configurable `ChunkedLoss` that is a memory-saving *executor* rather than a loss of its own: it owns the `lm_head` and runs a pluggable inner `loss_fn`
(a `BaseLoss`, defaulting to `CrossEntropyLoss`) over the model's hidden states one sequence chunk at a time. Every loss is unified under a single `__call__(pred, labels, global_valid_tokens, ...) -> (loss, metrics)` contract,
so callers never branch on the return type. A new logits-domain `GRPOLoss` (`BaseLoss`) is added that can be used standalone (model emits logits) or dropped inside `ChunkedLoss` (model skips its `lm_head`, the loss applies it per chunk), reusing the exact same chunking machinery as CE. TP loss-parallel is preserved for the chunked path via an explicit SPMD autograd `Function` (`_LossParallelCrossEntropy`) that replaces the DTensor `loss_parallel()` context manager. The change is behavior-preserving for SFT/CE (bitwise-identical to the legacy chunked CE) and selects the loss-parallel CE path purely by logits placement (vocab `Shard(-1)`), consistent with `main`'s flag-free design after the `--disable_loss_parallel` removal. All callsites (`torchtitan/trainer.py`, `validate.py`, Flux trainer/validate, forge `example_train.py`, graph_trainer, the PP schedule, and the RL policy trainer) are updated to the new `(loss, metrics)` contract; the RL GRPO trainer is rewired onto core `GRPOLoss`/`ChunkedLoss`.

## At a Glance

`ChunkedLoss` stops being a CE-specific loss and becomes a memory-saving *executor*: it owns the `lm_head` and runs a pluggable inner loss over hidden-state chunks, and every loss now returns the same `(loss, metrics)` pair.

```python
# main: a single, hard-coded chunked cross-entropy
ChunkedCELoss(num_chunks=8)

# this PR: ChunkedLoss wraps a *pluggable* inner loss (BaseLoss), same machinery
ChunkedLoss.Config(num_chunks=8)                              # default == legacy chunked CE
ChunkedLoss.Config(num_chunks=8, loss_fn=GRPOLoss.Config())  # chunked GRPO, no new plumbing

# ...and every loss now shares one calling contract:
loss, metrics = loss_fn(pred, labels, global_valid_tokens, **loss_inputs)
```

```mermaid
flowchart TD
    H["model forward, _skip_lm_head=True<br/>hidden states [B, L, D]"] --> S["detach + split along L into num_chunks"]
    S --> LOOP{{"for each chunk (inside spmd.local)"}}
    LOOP --> LM["lm_head(chunk) -> logits [B, L/N, V]<br/>vocab Shard(-1) under TP"]
    LM --> INNER["inner loss_fn(logits, labels, gvt, **loss_inputs)"]
    INNER -->|default| CE["CrossEntropyLoss<br/>(loss-parallel CE)"]
    INNER -->|RL| GRPO["GRPOLoss<br/>(per-token clipped surrogate)"]
    CE --> ACC["chunk.backward() -> GradAccumulator [B, L, D]"]
    GRPO --> ACC
    ACC --> LOOP
    LOOP -->|all chunks done| BRIDGE["_DecoderOutputGradientBackProp<br/>feeds accumulated grad into decoder backward"]
    BRIDGE --> OUT["return (loss, metrics)"]
```

Peak logits memory drops from `O(B*L*V)` to `O(B*L/num_chunks*V)`: the inner loss consumes logits one chunk at a time, so the full `[B, L, V]` tensor is never materialized across all chunks.

## Validation

- Core wandb curves, branch vs main: [wandb](https://api.wandb.ai/links/damian-szwichtenberg-intel/ucefmjam)
- RL GRPO chunked vs non-chunked, batch-invariant: [wandb](https://api.wandb.ai/links/damian-szwichtenberg-intel/hs1ihh72)

## For reviewers

There is loss difference between RL chunked and non-chunked - under verification.
Related RFC: #3477

### CODE REVIEW

File: torchtitan/experiments/rl/actors/trainer.py [hunk 1]
```diff
@@ -183,7 +156,19 @@ def __init__(
...
         self.config = config
         self.compile_config = compile_config
+        # Keep the loss uncompiled in RL: GRPOLoss has no compilable ``self.fn``
+        # and ChunkedLoss compiles only its inner leaf loss, so passing
+        # compile_config here would only emit a warning for GRPOLoss.
         self.loss_fn = config.loss.build()
+        # ChunkedLoss reduces logits to logprobs per chunk and never holds the
+        # full [B, L, V] logits, so the full-sequence logprob verification can't
+        # run for it. Decide once here (and warn) rather than per step.
+        self._verify_logprobs = not isinstance(self.loss_fn, ChunkedLoss)
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:166]: No let's not have this flag. If in practice we expect ChunkedLoss to be the default, this branch is meaningless.

File: torchtitan/experiments/rl/actors/trainer.py [hunk 2]
```diff
@@ -378,7 +373,29 @@ def reduce_forward_backward_metrics(
...
+            for key in keys:
+                value = values_by_key[key].detach()
+                if isinstance(value, DTensor):
+                    assert all(
+                        p.is_replicate() or p.is_partial() for p in value.placements
+                    ), (
+                        f"metric '{key}' is a DTensor with unsupported placements "
+                        f"{value.placements}; only Replicate/Partial are supported."
+                    )
+                    value = value.to_local()
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:396]: supervised training doesn't need this `to_local`, why it's needed here? https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -326,6 +342,190 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
         self._maybe_compile(compile_config)
 
 
+# Clamp |log(pi_theta / pi_old)| before exp() so a large generator/trainer
+# logprob mismatch -- notably the NaNs vLLM can emit under cudagraph -- cannot
+# overflow exp() to inf/NaN.
+_MAX_LOG_RATIO = 10.0
+
+
+class GRPOLoss(BaseLoss):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:351]: This class should be kept in rl/losses/ folder

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -326,6 +342,190 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
+                # (dropped above); tracked against the original response_mask.
+                "loss/generator_logprob_nan_frac": (
+                    (~torch.isfinite(generator_logprobs)).float() * response_mask
+                ).sum()
+                / loss_denominator,
+            }
+        return loss, metrics
+
+
+def logits_to_logprobs(logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:485]: `compute_logprobs` seems a conventional name, can we keep or do you have strong arguments for changing

File: torchtitan/components/loss.py [hunk 5]
```diff
@@ -284,11 +306,16 @@ def __call__(
         pred: torch.Tensor,
         labels: torch.Tensor,
         global_valid_tokens: float | None = None,
-    ) -> torch.Tensor:
+    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:309]: TBH I'm worried about the span of changes this PR creates, just for returning optional metrics (for RL only). Plus it can not compute metrics like logp diff between trainer and generator.

@felipemello1 what's your thought?

For now I'm OK with this signature returning a union of "loss | tuple of loss and metrics"

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -444,20 +644,43 @@ def result(self) -> torch.Tensor:
         return self._buffer
 
 
-class ChunkedCELoss(BaseLoss):
-    """Chunked cross-entropy loss that splits the sequence dimension to reduce peak memory.
+class ChunkedLoss(BaseLoss):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:647]: maybe explicit say this is a wrapper
```suggestion
class ChunkedLossWrapper(BaseLoss):
```

File: torchtitan/components/loss.py (hunk 4 above)
Comment (felipemello1) [reply torchtitan/components/loss.py:485]: Why don't we just call the cross_entropy_loss instead? Doesn't the SFT CE already take care of all the dtensor checks, etc? i dont see why we would reinvent it for GRPO

File: torchtitan/components/loss.py [hunk 7]
```diff
@@ -589,12 +832,20 @@ def _chunk(t):
...
+                # full-sequence result. With the default ``CrossEntropyLoss`` this is
+                # bitwise-identical to the legacy chunked cross-entropy.
+                chunk_inputs = {key: chunks[i] for key, chunks in input_chunks.items()}
+                chunk_loss, chunk_metrics = self.loss_fn(
+                    logits, label_chunk, global_valid_tokens, **chunk_inputs
                 )
-                if global_valid_tokens is not None:
-                    chunk_loss = chunk_loss / global_valid_tokens
+                for key, value in chunk_metrics.items():
+                    metrics[key] = metrics.get(key, 0) + value
```
Comment (felipemello1) [inline torchtitan/components/loss.py:847]: i dont think that you should sum them. What if the metric is "max"/"min" or something else?

For now, we could have a simple util that checks if the metric ends in /mean, /frac, /max, /min, and acts accordingly. Thats what we do in experiments/rl/trainer.py.

In the future we will have some proper support for distributed metrics

File: torchtitan/components/loss.py [hunk 8]
```diff
@@ -609,14 +860,20 @@ def _chunk(t):
                 lm_head.set_reshard_after_backward(True)
                 lm_head.set_requires_gradient_sync(True, recurse=False)
                 lm_head.reshard()
+
+            if metrics:
+                metrics["loss/mean"] = total_loss.detach()
```
Comment (felipemello1) [inline torchtitan/components/loss.py:865]: The loss should take care of this, not the wrapper

File: torchtitan/components/loss.py (hunk 5 above)
Comment (felipemello1) [reply torchtitan/components/loss.py:309]: >  Plus it can not compute metrics like logp diff between trainer and generator.

hmmm, we should be able to move this metric inside of the GPRO loss, no? I think it would be ideal.

> TBH I'm worried about the span of changes this PR creates, just for returning optional metrics (for RL only)

I skimmed over the loss.py but i didn't see too many changes because of it. Are you referring to other files?

Do you think that the z-loss metric would benefit from it? i didnt review the PR that added it, so i dont know.

>For now I'm OK with this signature returning a union of "loss | tuple of loss and metrics"

Noob question: wouldn't the wrapper still need to be aware of it anyway? Its not clear to me which functions would be simplified because of that. The call sites still need to be updated anyway because of the `global_valid_tokens` arg change :/

Image

Comment (felipemello1) [reply torchtitan/components/loss.py:309]: I think its ok to change to "loss | tuple of loss and metrics", but its not clear to me what it simplifies. I am afraid that it would do the opposite, because now the caller has to check if its a tuple

**Comment by felipemello1** [comment]
hey @DamianSzwichtenberg , i see that its still marked as a draft, but could you update the pr description with some code snippet or diagram that is more human friendly to get the gist of the changes?

File: torchtitan/experiments/rl/actors/trainer.py [hunk 9]
```diff
@@ -430,41 +447,57 @@ async def forward_backward(
...
+        loss_metrics["loss/mean"] = loss.detach()
+
+        sum_reduced_metrics = dict(loss_metrics)
+        max_reduced_metrics: dict[str, torch.Tensor] = {}
+
+        # Logprob identity verification needs the full-sequence policy logprobs.
+        # The non-chunked path still has the logits in scope, so recompute the
+        # logprobs cheaply under no_grad. The chunked path never materializes the
+        # full logits, so verification is skipped (decided in __init__).
+        if self._verify_logprobs:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:482]: also, if we return the logprobs of shape [B, L] concatenated from chunked loss forward, we could still compute the logp diff

**Comment by tianyu-l** [comment]
@claude please review this PR

**Comment by DamianSzwichtenberg** [comment]
Hi all, quick heads-up: I'm on vacation until June 29 and won't have access to my computer in the meantime. I'll address all the comments once I'm back.

Feel free to push modifications to the PR while I'm away, I'm fine with that.

Thanks for your patience!

**Comment by felipemello1** [comment]
hey @DamianSzwichtenberg , thanks for kickstarting this! We had a deadline, so pianpwk finished things off for us. Looking forward for more PRs from you in the future if you are interested. Hope you had a good time off.

image
